src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/FindAllReferences/RemoteFindAllReferencesService.cs

```diff
@@ -0,0 +1,116 @@
+{
+    internal sealed class Factory : FactoryBase<IRemoteFindAllReferencesService>
+    {
+        protected override IRemoteFindAllReferencesService CreateService(in ServiceArgs args)
+            => new RemoteFindAllReferencesService(in args);
+    }
+
+    private readonly IClientCapabilitiesService _clientCapabilitiesService = args.ExportProvider.GetExportedValue<IClientCapabilitiesService>();
+
+    protected override IDocumentPositionInfoStrategy DocumentPositionInfoStrategy => PreferAttributeNameDocumentPositionInfoStrategy.Instance;
+
+    public ValueTask<RemoteResponse<SumType<VSInternalReferenceItem, LspLocation>[]?>> FindAllReferencesAsync(
+        JsonSerializableRazorPinnedSolutionInfoWrapper solutionInfo,
+        JsonSerializableDocumentId documentId,
+        Position position,
+        CancellationToken cancellationToken)
+        => RunServiceAsync(
+            solutionInfo,
+            documentId,
+            context => GetImplementationAsync(context, position, cancellationToken),
+            cancellationToken);
+
+    private async ValueTask<RemoteResponse<SumType<VSInternalReferenceItem, LspLocation>[]?>> GetImplementationAsync(
+        RemoteDocumentContext context,
+        Position position,
+        CancellationToken cancellationToken)
+    {
+        var codeDocument = await context.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+
+        if (!codeDocument.Source.Text.TryGetAbsoluteIndex(position, out var hostDocumentIndex))
+        {
+            return NoFurtherHandling;
+        }
+
+        var positionInfo = GetPositionInfo(codeDocument, hostDocumentIndex, preferCSharpOverHtml: true);
+
+        if (positionInfo.LanguageKind is not RazorLanguageKind.CSharp)
+        {
+            return NoFurtherHandling;
+        }
+
+        // Finally, call into C#.
+        var generatedDocument = await context.Snapshot
+            .GetGeneratedDocumentAsync(cancellationToken)
+            .ConfigureAwait(false);
+
+        var results = await ExternalHandlers.FindAllReferences
+            .FindReferencesAsync(
+                RemoteWorkspaceAccessor.GetWorkspace(),
+                generatedDocument,
+                VsLspExtensions.ToLinePosition(positionInfo.Position),
+                _clientCapabilitiesService.ClientCapabilities.SupportsVisualStudioExtensions,
+                cancellationToken)
+            .ConfigureAwait(false);
+
+        if (results is null and not [])
+        {
+            // C# didn't return anything, so we're done.
+            return NoFurtherHandling;
+        }
+
+        // Map the C# locations back to the Razor file.
+        foreach (var result in results)
+        {
+            var location = result.TryGetFirst(out var referenceItem)
+                ? referenceItem.Location
+                : result.Second;
+
+            if (location is null)
```

Comment:
Honestly, I've no idea, but `referenceItem.Location` is annotated as nullable so this defensiveness is just to make the compiler happy. I suspect this falls out as a quirk in the VS LSP client, where the underlying VS type doesn't have to specify a location, because navigation can be done by a callback, but callbacks aren't part of LSP.